### PR TITLE
MLperf BERT training

### PR DESCRIPTION
**Pull request by Eliulm**
### Step 1: Downloading, Verifying, and Uncompressing Datasets
Ensure the dataset is properly downloaded, verified against an MD5 hash, and uncompressed to your specified base directory.

**Command:**
```shell
VERIFY=1 TRAIN_DOWNLOAD=1 BASEDIR=/raid/datasets/wiki python3 extra/datasets/wikipedia_download.py
```

- `VERIFY`: Set to 1 to enable verification against MD5 hash.
- `TRAIN_DOWNLOAD`: Set to 1 to initiate download of the training dataset.
- `BASEDIR`: Specify the directory to store the dataset.

### Step 2: Offline Data Preprocessing

**Preprocess All Train Data Files:**

- Preprocesses all 500 train data files. Note: Using more than 32 threads leads to RAM shortages.

  **Command:**
  ```shell
  NUM_WORKERS=32 BASEDIR=/raid/datasets/wiki python3 extra/datasets/wikipedia.py pre-train all
  ```

     - `NUM_WORKERS`: Maximum number of threads to use.
**Preprocess a Specific Train Data File (0-499):**

  **Command:**
  ```shell
  NUM_WORKERS=32 BASEDIR=/raid/datasets/wiki python3 extra/datasets/wikipedia.py pre-train 0
  ```

**Preprocess the Validation Dataset (10 Files):**

  **Command:**
  ```shell
  BASEDIR=/raid/datasets/wiki python3 extra/datasets/wikipedia.py pre-eval
  ```

### Step 3: Training on TinyBox

Set the relevant environment variables for training based on your setup.

#### Learning Rate Related Variables:

- `GLOBAL_BATCH_SIZE`: The device batch size. For Tinybox with 6 GPUs in FP32 `24`, in FP16 `<fill>`.
- `EVAL_BS`: Evaluation batch size. For Tinybox, use `12`.
- `OPT_BASE_LEARNING_RATE`: The maximum learning rate. For Tinybox: `0.0001`.
- `TRAIN_STEPS`: Total number of training steps per epoch. For Tinybox: `150000`. 
- `NUM_WARMUP_STEPS`: Number of warmup steps. Generally, 10% of `TRAIN_STEPS`.

#### Performance Related Variables:

- `BEAM`: Performance setting, with diminishing returns after `BEAM=4`. For Tinybox, use `4`.
- `DEFAULT_FLOAT`: Specify as `half` or `bfloat16`. Note: Both do not work yet.
- `LOSS_SCALER`: Value to adjust loss values into the representable range of FP16.
- `GPUS`: Number of GPUs. For Tinybox, use `6`.

#### Miscellaneous Variables:

- `WANDB`: Set to enable Weights & Biases logging.
- `CKPT`: Set to enable checkpointing.
- `CKPT_DIR`: Directory to save checkpoints. For Tinybox, use `/raid/ckpts`.
- `MODEL`: `bert`
- `BASEDIR`: Directory where data and Bert configuration files are saved. For Tinybox, use `/raid/datasets/wiki`.

**Full Command for FP32 Training on TinyBox with Recommended Presets:**
```shell
GLOBAL_BATCH_SIZE=24 EVAL_BS=12 OPT_BASE_LEARNING_RATE=0.0001 TRAIN_STEPS=150000 NUM_WARMUP_STEPS=15000 BEAM=4 GPUS=6 WANDB=1 CKPT=1 CKPT_DIR=/raid/ckpts MODEL=bert python3 examples/mlperf/model_train.py
```
`<Fill>`: will be determined soon.

### CODE REVIEW

**Comment by geohot** [comment]
What's your discord? I'll add you to the "#tinybox-access" channel

**Comment by Eliulm** [comment]
Display name is `eliulm` and username is `eliaaaaaaas`.

**Comment by Eliulm** [comment]
@geohot Update: I am seeing a lot of python time with BERT training: `2  932.44 ms run,    1.22 ms fetch,  475.75 ms python,  455.47 ms CUDA,   10.90 loss, 0.000002 LR, 13.68 GB used,   6949.30 GFLOPS`.

Currently doing [runs](https://wandb.ai/wahl/MLPerf-BERT?nw=nwuserwahl) test more aggressive LR schedules. Trying to get the training time down to 2-3 days on my 4080.

**Comment by chenyuxyz** [comment]
Is there a command that I can run to get the data, verify the data, and train the model? I am interested in testing this on a tinybox.

**Comment by Eliulm** [comment]
@chenyuxyz, here are the steps:

1. **Download and Uncompress:**
   ```bash
   TRAIN_DOWNLOAD=1 VERIFY=1 python3 extra/datasets/wikipedia_download.py
   ```

2. **Preprocessing for Training:**
   - To preprocess and train on all samples, use:
     ```bash
     NUM_WORKERS=32 python3 extra/datasets/wikipedia.py pre-train all
     ```
   - To preprocess and train on a specific set (e.g., 0-500, which will have about 700,000 data samples each):
     ```bash
     python3 extra/datasets/wikipedia.py pre-train 0
     ```

3. **Preprocessing for Evaluation:**
   ```bash
   python3 extra/datasets/wikipedia.py pre-eval
   ```

4. **Training:**
   - Note: Using `HALF=1` currently results in `NaN` losses.
   - For batch size, I have not yet tested with the new scheduler.
     ```bash
     python3 examples/mlperf/model_train.py

**Comment by chenyuxyz** [comment]
is the processed data somewhere on tiny10? I would like to just to copy data to other tinybox instead of rerunning the processing job

**Comment by Eliulm** [comment]
@chenyuxyz Unfortunately, no. Only a few sets on tiny10.

**Comment by chenyuxyz** [comment]
fyi generating all training data will create too many files and eventually use all inodes

**Comment by Eliulm** [comment]
Yeah, I will write multiple instances to a file. Wanted to refactor the dataloaders anyway.

**Comment by Eliulm** [comment]
@chenyuxyz Datasamples are now saved in batches of 1000 to a file, and I finished refactoring the training script. To be 100% MLperf compliant, I have to do a few minor tweaks, but most of the work should be done. Also, since my last rebase, I noticed that training performance has more than halved. I will try to isolate the specific commit that caused the performance regression and see what I can do.

**Comment by chenyuxyz** [comment]
you can use `DEFAULT_FLOAT=` to specify the default float
